server/src/main/java/org/opensearch/index/remote/RemoteIndexPathUploader.java

```diff
@@ -0,0 +1,267 @@
+                );
+                actionListener.onFailure(ex);
+                return;
+            }
+            if (exceptionList.size() > 0) {
+                RemoteStateTransferException ex = new RemoteStateTransferException(
+                    String.format(Locale.ROOT, UPLOAD_EXCEPTION_MSG, indexNames)
+                );
+                exceptionList.forEach(ex::addSuppressed);
+                actionListener.onFailure(ex);
+                return;
+            }
+            success = true;
+            actionListener.onResponse(null);
+        } finally {
+            long tookTimeNs = System.nanoTime() - startTime;
+            logger.trace(new ParameterizedMessage("executed beforeNewIndexUpload status={} tookTimeNs={}", success, tookTimeNs));
+        }
+
+    }
+
+    private void writeIndexPathAsync(IndexMetadata idxMD, CountDownLatch latch, List<Exception> exceptionList) throws IOException {
+        Map<String, String> remoteCustomData = idxMD.getCustomData(IndexMetadata.REMOTE_STORE_CUSTOM_KEY);
+        RemoteStoreEnums.PathType pathType = RemoteStoreEnums.PathType.valueOf(remoteCustomData.get(RemoteStoreEnums.PathType.NAME));
+        RemoteStoreEnums.PathHashAlgorithm hashAlgorithm = RemoteStoreEnums.PathHashAlgorithm.valueOf(
+            remoteCustomData.get(RemoteStoreEnums.PathHashAlgorithm.NAME)
+        );
+        String indexUUID = idxMD.getIndexUUID();
+        int shardCount = idxMD.getNumberOfShards();
+        BlobPath translogBasePath = translogRepository.basePath();
+        BlobContainer translogBlobContainer = translogRepository.blobStore().blobContainer(translogBasePath.add(RemoteIndexPath.DIR));
+
+        if (isTranslogSegmentRepoSame) {
+            // If the repositories are same, then we need to upload a single file containing paths for both translog and segments.
+            Map<RemoteStoreEnums.DataCategory, List<RemoteStoreEnums.DataType>> pathCreationMap = new HashMap<>();
+            pathCreationMap.putAll(TRANSLOG_PATH);
+            pathCreationMap.putAll(SEGMENT_PATH);
+            REMOTE_INDEX_PATH_FORMAT.writeAsyncWithUrgentPriority(
+                new RemoteIndexPath(indexUUID, shardCount, translogBasePath, pathType, hashAlgorithm, pathCreationMap),
+                translogBlobContainer,
+                indexUUID,
+                getUploadPathLatchedActionListener(idxMD, latch, exceptionList, pathCreationMap)
+            );
+        } else {
+            // If the repositories are different, then we need to upload one file per segment and translog containing their individual
+            // paths.
+            REMOTE_INDEX_PATH_FORMAT.writeAsyncWithUrgentPriority(
+                new RemoteIndexPath(indexUUID, shardCount, translogBasePath, pathType, hashAlgorithm, TRANSLOG_PATH),
+                translogBlobContainer,
+                indexUUID,
+                getUploadPathLatchedActionListener(idxMD, latch, exceptionList, TRANSLOG_PATH)
+            );
```

Comment:
seems like repeated code 3 time, can this be moved to private method?

Reply:
Done.